### PR TITLE
Add `metrics` parameter to `data_tabulate()`

### LINKED ISSUE #685: Cumulative distribution should be optional
Thanks so much for making a useful frequency and cross tabs system! 

Right now on the frequency tables the cumulative distribution is always included. However, in many/most 
cases--meaning when variables are not ordinal--cumulative distribution is both irrelevant and confusing. I'm getting ready to teach this next week and I had to write in the documentation "ignore those columns for now." 

It looks like it would be pretty easy to add a parameter to make the cumulative conditional or to only include those columns when `is.ordered()` or `is.numeric` return `TRUE`. 

https://github.com/easystats/datawizard/blob/main/R/data_tabulate.R#L262

I'd be happy to take a swing at implementing if it is of interest.

### PR DESCRIPTION
**Pull request by elinw**
There were a few ways to approach this. The cumulative values depends on having the valid values, and it could make sense to just set any not included measures to NULL at the end, the way I did for valid.

I added a test for `measures = "raw"` ... I could add for other variations also.

Fixes #685

### CODE REVIEW

**Comment by elinw** [comment]
#685

File: R/data_tabulate.R [hunk 1]
```diff
@@ -154,6 +158,7 @@ data_tabulate.default <- function(
   proportions = NULL,
   name = NULL,
   verbose = TRUE,
+  measures = c("raw", "valid", "cumulative"),
```
Comment (strengejacke) [inline R/data_tabulate.R]: Can you validate the input, e.g. using `match.arg(..., several.ok = TRUE)`?

File: R/data_tabulate.R [hunk 2]
```diff
@@ -250,7 +255,9 @@ data_tabulate.default <- function(
     out$N <- round(out$N)
   }
 
-  out$`Raw %` <- 100 * out$N / sum(out$N)
+  if ("raw" %in% measures){
```
Comment (strengejacke) [inline R/data_tabulate.R]: Please use space around infix-operators :-)

File: tests/testthat/test-data_tabulate.R [hunk 3]
```diff
@@ -181,6 +181,73 @@ test_that("data_tabulate data.frame", {
...
+        unique(efc$e16sex)
+      ),
+      NA
+    ))
+  )
+  expect_identical(table1$N, as.vector(table(addNA(efc$e16sex))))
+  expect_identical(
+    table1$`Raw %`,
+    as.vector(c(
+      100 * table(efc$e16sex, useNA = "always") / length(efc$e16sex)
```
Comment (strengejacke) [inline tests/testthat/test-data_tabulate.R:242]: See lintr, you may remove `c()` here.

File: R/data_tabulate.R [hunk 4]
```diff
@@ -39,6 +39,10 @@
 #' printed as markdown or HTML table, depending on the environment. See
 #' [`insight::export_table()`] for details.
 #' @param verbose Toggle warnings and messages.
+#' @param measures Optional character vector, indicating the types of
+#' percents to be included. Only applies to frequencies, i.e. when `by` is
+#' `NULL`. Can be `"raw"` (includes `NA` values), `"valid"` (excludes `NA` values)
+#' or `"cumulative"` (excludes `NA` vlues).
```
Comment (strengejacke) [inline R/data_tabulate.R]: Typo (`vlues`)

**Review by etiennebacher — CHANGES_REQUESTED** [review]
On top of the other comments to address above, please add a bullet point in `NEWS.md` and add an example for the new argument in the docs of `data_tabulate()`.

(@strengejacke I hid all the copilot comments because they were just duplicates of your comments and adding noise to this PR)

File: R/data_tabulate.R (hunk 1 above)
Comment (elinw) [reply R/data_tabulate.R]: Oh missed this, will do this now.

Comment (elinw) [reply R/data_tabulate.R]: Is it okay that it returns an error if none of the values given are valid?

**Comment by elinw** [comment]
I was wondering if it should be documented that `measures = c()` will return the frequencies only.
Also, that made me wonder if "N" should also be optional.

File: R/data_tabulate.R (hunk 2 above)
Comment (elinw) [reply R/data_tabulate.R]: This is fixed, not sure why it doesn't show as outdated.

File: R/data_tabulate.R [hunk 5]
```diff
@@ -86,6 +90,9 @@
 #' # drop missing values
 #' data_tabulate(efc$c172code, remove_na = TRUE)
 #'
+#' # exclude the cumulative percent column
+#' data_tabulate(efc$c172code, remove_na = TRUE)
+#'
```
Comment (etiennebacher) [inline R/data_tabulate.R:96]: The new example should show how to use `measures` instead of `remove_na`.

File: R/data_tabulate.R [hunk 6]
```diff
@@ -39,6 +39,10 @@
 #' printed as markdown or HTML table, depending on the environment. See
 #' [`insight::export_table()`] for details.
 #' @param verbose Toggle warnings and messages.
+#' @param measures Optional character vector, indicating the types of
+#' percents to be included. Only applies to frequencies, i.e. when `by` is
+#' `NULL`. Can be `"raw"` (includes `NA` values), `"valid"` (excludes `NA` values)
+#' or `"cumulative"` (excludes `NA` values).
```
Comment (etiennebacher) [inline R/data_tabulate.R]:
```suggestion
#' `NULL`. Can be several choices among `"raw"` (includes `NA` values), 
#' `"valid"` (excludes `NA` values) and `"cumulative"` (excludes `NA` values).
```

File: tests/testthat/test-data_tabulate.R [hunk 7]
```diff
@@ -181,6 +181,73 @@ test_that("data_tabulate data.frame", {
   )
 })
 
+test_that("data_tabulate data.frame with measures", {
```
Comment (etiennebacher) [inline tests/testthat/test-data_tabulate.R]: Please add tests for `measures = c()`, `measures = NULL` and `measures = "foo"`.

File: NEWS.md [hunk 8]
```diff
@@ -20,6 +20,9 @@ CHANGES
   `.rds`, and `.rdata`). Data encryption is based on the AES-GCM algorithm using
   the `openssl::aes_gcm_encrypt()` function (#675).
 
+* `data_tabulate()` gain a `measures` argument to allow selection of columns to
+  display ("raw", "valid", and "cumulative") (#689).
```
Comment (etiennebacher) [inline NEWS.md]: You can add your github username if you want: `(#689, <username>)`

**Review by etiennebacher — COMMENTED** [review]
> I was wondering if it should be documented that measures = c() will return the frequencies only.

Yes I think so.

>  Also, that made me wonder if "N" should also be optional.

I don't know about that but if it's not included then I think this `measures` parameter should be renamed `shares`. @strengejacke WDYT?

---

The workflow for code styling fails. We use [Air](https://posit-dev.github.io/air/) to automatically format code so you can install it and run it on the project to fix this failure.

**Comment by elinw** [comment]
> > I was wondering if it should be documented that measures = c() will return the frequencies only.
> 
> Yes I think so.
> 
> > Also, that made me wonder if "N" should also be optional.
> 
> I don't know about that but if it's not included then I think this `measures` parameter should be renamed `shares`. @strengejacke WDYT?
> 
> The workflow for code styling fails. We use [Air](https://posit-dev.github.io/air/) to automatically format code so you can install it and run it on the project to fix this failure.

"shares" is not meaningful to me.  Maybe "percentages" would be more descriptive?

File: tests/testthat/test-data_tabulate.R (hunk 7 above)
Comment (elinw) [reply tests/testthat/test-data_tabulate.R]: What do you think should happen for measures = NULL?  In the current design it acts like c(), which makes sense to me. However, just like with measures = "foo" it ignores without erroring.

Tests are now added on that basis.

File: R/data_tabulate.R (hunk 6 above)
Comment (elinw) [reply R/data_tabulate.R]: I changed the wording but not exactly to yours. See what you think.

**Comment by strengejacke** [comment]
If `N` is included, `percentages` would not fit perfectly. I would say `metrics` (if `N` should be included) or `percentages` is fine.

**Comment by elinw** [comment]
> If `N` is included, `percentages` would not fit perfectly. I would say `metrics` (if `N` should be included) or `percentages` is fine.

This sounds good.  Can you let me know if `N` definitely should be included? 
I'll do a commit for the name change alone first.

**Comment by elinw** [comment]
> > If `N` is included, `percentages` would not fit perfectly. I would say `metrics` (if `N` should be included) or `percentages` is fine.
> 
> This sounds good. Can you let me know if `N` definitely should be included? I'll do a commit for the name change alone first.

I added the "N" (in uppercase) option. I did have to change one additional function that assumed that the N column would be present.  All of the tests still pass.

On the tests, the linter does not like testing for `metrics = c()`.  I do think people will write that exactly so it's good to check, but `identical(c(), NULL)` is `TRUE` so technically it's a redundant test.  Let me know what you think.

**Comment by elinw** [comment]
I think everything is resolved unless you have new requests.

File: R/data_tabulate.R [hunk 9]
```diff
@@ -39,6 +39,12 @@
 #' printed as markdown or HTML table, depending on the environment. See
 #' [`insight::export_table()`] for details.
 #' @param verbose Toggle warnings and messages.
+#' @param metrics Optional character vector, indicating the types of
+#' percents to be included. Only applies to frequencies, i.e. when `by` is
+#' `NULL`. Can include any combination of `N` (frequencies including NA), `"raw"` (includes `NA` values),
+#' `"valid"` (excludes `NA` values) or `"cumulative"` (excludes `NA` values).
+#' Using `c()` will return a table with only the value labels. Invalid
+#' values (`metrics = "foo"`) are silently ignored.
```
Comment (etiennebacher) [inline R/data_tabulate.R]:
```suggestion
#' @param metrics Optional character vector, indicating the types of
#' metrics to be included. Only applies to frequencies, i.e. when `by` is
#' `NULL`. Can include any combination of `N` (frequencies including `NA`), 
#' `"raw"` (percentage including `NA` values), `"valid"` (percentage excluding
#' `NA` values) or `"cumulative"` (percentage excluding `NA` values).
```
The part about silently ignoring unknown values seemed wrong (it errors) so I removed it.

File: R/data_tabulate.R [hunk 10]
```diff
@@ -86,6 +92,12 @@
 #' # drop missing values
 #' data_tabulate(efc$c172code, remove_na = TRUE)
 #'
+#' # exclude the cumulative percent column
+#' data_tabulate(efc$c172code, metrics = c("raw", "valid"))
+#'
+#' # return frequencies only
+#' data_tabulate(efc$c172code, remove_na = TRUE, metrics = c())
```
Comment (etiennebacher) [inline R/data_tabulate.R]: The comment is outdated but I don't think we should showcase the `metrics = c()` anyway, it's just an edge case. I think these two lines can be removed.

File: R/data_tabulate.R [hunk 11]
```diff
@@ -276,6 +304,10 @@ data_tabulate.default <- function(
     }
     out <- cbind(var_info, out)
   }
+  total_n <- sum(out$N, na.rm = TRUE)
```
Comment (etiennebacher) [inline R/data_tabulate.R:317]: Why has this line moved? I don't think it causes an issue in itself but I'd rather minimize the git diff.

File: NEWS.md [hunk 12]
```diff
@@ -20,6 +20,9 @@ CHANGES
   `.rds`, and `.rdata`). Data encryption is based on the AES-GCM algorithm using
   the `openssl::aes_gcm_encrypt()` function (#675).
 
+* `data_tabulate()` gain a `measures` argument to allow selection of columns to
+  display ("raw", "valid", and "cumulative") (#689, @elinw).
```
Comment (etiennebacher) [inline NEWS.md]:
```suggestion
* `data_tabulate()` gain a `metrics` argument to allow selection of columns to
  display (#689, @elinw).
```

**Review by etiennebacher — CHANGES_REQUESTED** [review]

**Comment by etiennebacher** [comment]
Can you also resolve the conflicts?

**Comment by elinw** [comment]
I tried running Air on the file but the only issue it identified was duplicate blank lines. Not sure why this is failing.

**Comment by etiennebacher** [comment]
The problem was that a new version of Air was released a couple of days ago and replaces `=` by `<-` by default (when possible). This was clashing with our custom syntax that can be used in `data_filter()`.

There are just the few comments above to resolve now (don't worry about linting failures).

**Comment by elinw** [comment]
> The problem was that a new version of Air was released a couple of days ago and replaces `=` by `<-` by default (when possible). This was clashing with our custom syntax that can be used in `data_filter()`.
> 
> There are just the few comments above to resolve now (don't worry about linting failures).

How about having a separate branch for those, since the comments are in unrelated files?

**Comment by etiennebacher** [comment]
Yes I have actually merged a PR that fixes those Air failures so that these changes don't appear here. Your latest commit message says you left questions on some comments but I don't see anything, am I missing something? I just want to avoid a situation where we're both waiting for something from the other.

**Comment by elinw** [comment]
> Yes I have actually merged a PR that fixes those Air failures so that these changes don't appear here. Your latest commit message says you left questions on some comments but I don't see anything, am I missing something? I just want to avoid a situation where we're both waiting for something from the other.

I think, in the end, the only question I had was about metrics = "foo" and that doesn't matter.

**Comment by etiennebacher** [comment]
There are still unresolved comments above about docs/NEWS/line changed. I have made "suggestions" using Github interface but didn't actually apply them.

**Comment by elinw** [comment]
I'm not sure why two builds are failing, but they don't seem related. R crashed during the test of ` test-standardize_models.R: ` but it might be that retriggering the builds would fix. Not sure what would have changed. 
R 4.6.1 was released today so I guess it couldbe related to that.

Update:

── Warning (test-standardize-data.R:23:1): standardize.numeric ─────────────────
Global state has changed:
     names(before[[4]]) | names(after[[4]])      
[79] "TESTTHAT"         | "TESTTHAT"         [79]
[80] "TESTTHAT_WD"      | "TESTTHAT_WD"      [80]
[81] "TMPDIR"           | "TMPDIR"           [81]
                        - "TZDIR"            [82]
[82] "USER"             | "USER"             [83]
[83] "XPC_FLAGS"        | "XPC_FLAGS"        [84]
[84] "XPC_SERVICE_NAME" | "XPC_SERVICE_NAME" [85]
before[[4]][79:84] vs after[[4]][79:85]
  TERM"xterm-256color"
  TESTTHAT"true"
  TESTTHAT_WD"/Users/elinwaring/Code/rcode/datawizard"
  TMPDIR"/var/folders/gw/_qnrkpdn3fjg_0gwwnb9h_g40000gn/T/"
+ TZDIR"/var/db/timezone/zoneinfo"
  USER"elinwaring"
  XPC_FLAGS"0x0"
  XPC_SERVICE_NAME"application.com.rstudio.desktop.371427447.371427452"                                                                                                                                                             
[ FAIL 0 | WARN 1 | SKIP 0 | PASS 75 ]

update 2

https://discourse.mc-stan.org/t/global-state-has-changed-warning/36591

Was this resolved for `report`?  

update 3

Apparently the resolved this by deleting `tests/testthat/helper-state.R`

https://github.com/easystats/report/pull/471

**Comment by etiennebacher** [comment]
Failure on R 4.1 comes from https://github.com/mitchelloharawild/distributional/issues/157

**Review by etiennebacher — APPROVED** [review]
Thanks!
